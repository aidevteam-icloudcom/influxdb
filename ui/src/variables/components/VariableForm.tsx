// Libraries
import React, {PureComponent, ChangeEvent} from 'react'

// Components
import {Form, Input, Button, Grid} from '@influxdata/clockface'
import FluxEditor from 'src/shared/components/FluxEditor'

// Utils
import {validateVariableName} from 'src/variables/utils/validation'

// Types
import {IVariable as Variable} from '@influxdata/influx'
import {ButtonType, ComponentColor} from '@influxdata/clockface'

interface Props {
  onCreateVariable: (variable: Pick<Variable, 'name' | 'arguments'>) => void
  onHideOverlay?: () => void
  initialScript?: string
  variables: Variable[]
}

interface State {
  name: string
  script: string
}

export default class VariableForm extends PureComponent<Props, State> {
  constructor(props) {
    super(props)
    this.state = {
      name: '',
      script: this.props.initialScript || '',
    }
  }

  public render() {
    const {onHideOverlay} = this.props
    const {name, script} = this.state

    return (
      <Form onSubmit={this.handleSubmit}>
        <Grid>
          <Grid.Row>
            <Grid.Column>
              <div className="overlay-flux-editor--spacing">
                <Form.ValidationElement
                  label="Name"
                  value={name}
                  required={true}
                  validationFunc={this.handleNameValidation}
                >
                  {status => (
                    <Input
                      placeholder="Give your variable a name"
                      name="name"
                      autoFocus={true}
                      value={name}
                      onChange={this.handleChangeInput}
                      status={status}
                    />
                  )}
                </Form.ValidationElement>
              </div>
            </Grid.Column>
            <Grid.Column>
              <Form.Element label="Value">
                <div className="overlay-flux-editor">
                  <FluxEditor
                    script={script}
                    onChangeScript={this.handleChangeScript}
                    visibility="visible"
                    suggestions={[]}
                  />
                </div>
              </Form.Element>
            </Grid.Column>
            <Grid.Column>
              <Form.Footer>
                <Button
                  text="Cancel"
                  color={ComponentColor.Danger}
                  onClick={onHideOverlay}
                />
                <Button
                  text="Create"
                  type={ButtonType.Submit}
                  color={ComponentColor.Primary}
                />
              </Form.Footer>
            </Grid.Column>
          </Grid.Row>
        </Grid>
      </Form>
    )
  }

  private handleSubmit = (): void => {
    const {onCreateVariable, onHideOverlay} = this.props

    onCreateVariable({
      name: this.state.name,
      arguments: {
        type: 'query',
        values: {query: this.state.script, language: 'flux'},
      },
    })

    onHideOverlay()
  }

  private handleNameValidation = (name: string) => {
    const {variables} = this.props

    return validateVariableName(name, variables).error
  }

  private handleChangeInput = (e: ChangeEvent<HTMLInputElement>) => {
    const {value, name} = e.target

    const newState = {...this.state}
    newState[name] = value
    this.setState(newState)
  }

  private handleChangeScript = (script: string): void => {
    this.setState({script})
  }
}
