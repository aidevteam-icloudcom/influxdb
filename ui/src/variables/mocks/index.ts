// Types
import {Variable} from '@influxdata/influx'

export const createVariable = (
  name: string,
  query: string,
  selected?: string
): Variable => ({
  name,
  id: name,
  orgID: 'howdy',
  selected: selected ? [selected] : [],
  arguments: {
    type: 'query',
    values: {
      query,
      language: 'flux',
    },
  },
})
