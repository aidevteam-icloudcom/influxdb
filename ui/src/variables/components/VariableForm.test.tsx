// Libraries
import React from 'react'
import {shallow} from 'enzyme'

// Components
import VariableForm from 'src/variables/components/VariableForm'

// Constants
import {variables} from 'mocks/dummyData'

const setup = (override?) => {
  const actions = {
    name: jest.fn(),
    type: jest.fn(),
    query: jest.fn(),
    map: jest.fn(),
    constant: jest.fn(),
  }
  const props = {
    variables,
    variableType: 'query',
    query: {
      type: 'query',
      values: {
        query: '',
        language: 'flux',
      },
    },
    map: {
      type: 'map',
      values: {},
    },
    constant: {
      type: 'constant',
      values: [],
    },
    initialScript: 'Hello There!',
    onCreateVariable: jest.fn(),
    onHideOverlay: jest.fn(),
    onNameUpdate: name => actions.name(name),
    onTypeUpdate: type => actions.type(type),
    onQueryUpdate: arg => actions.query(arg),
    onMapUpdate: arg => actions.map(arg),
    onConstantUpdate: arg => actions.constant(arg),
    ...override,
  }

  const wrapper = shallow<VariableForm>(<VariableForm {...props} />)

  return {wrapper, actions}
}

describe('VariableForm', () => {
  describe('rendering', () => {
    it('renders', () => {
      const {wrapper} = setup()
      expect(wrapper.exists()).toBe(true)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('type selector', () => {
    it('should not tell the the parrent if the type was not changed', () => {
      const {wrapper, actions} = setup()

      const defaultType = 'query'

      // this way of accessing the instance function
      // bypasses typescript's private function check
      wrapper.instance()['handleChangeType'](defaultType)
      expect(actions.type.mock.calls.length).toBe(0)
    })

    it('should update tell the the parrent if type was changed', () => {
      const {wrapper, actions} = setup()

      // this way of accessing the instance function
      // bypasses typescript's private function check
      wrapper.instance()['handleChangeType']('map')
      expect(actions.type.mock.calls.length).toBe(1)
      expect(actions.type.mock.calls[0][0]).toBe('map')
    })
  })
})
