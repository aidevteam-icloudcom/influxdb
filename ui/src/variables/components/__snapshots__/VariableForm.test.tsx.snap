// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VariableForm rendering renders 1`] = `
<Form
  onSubmit={[Function]}
  preventDefault={true}
  testID="form-container"
>
  <Grid
    testID="grid"
  >
    <GridRow
      testID="grid--row"
    >
      <GridColumn
        testID="grid--column"
        widthXS={6}
      >
        <div
          className="overlay-flux-editor--spacing"
        >
          <FormValidationElement
            label="Name"
            required={true}
            testID="form--validation-element"
            validationFunc={[Function]}
          >
            <Component />
          </FormValidationElement>
        </div>
      </GridColumn>
      <GridColumn
        testID="grid--column"
        widthXS={6}
      >
        <FormElement
          label="Type"
          required={true}
          testID="form--element"
        >
          <Dropdown
            button={[Function]}
            dropUp={false}
            menu={[Function]}
            testID="dropdown"
          />
        </FormElement>
      </GridColumn>
    </GridRow>
    <GridRow
      testID="grid--row"
    >
      <GridColumn
        testID="grid--column"
        widthXS={12}
      >
        <VariableArgumentsEditor
          args={
            Object {
              "type": "query",
              "values": Object {
                "language": "flux",
                "query": "Hello There!",
              },
            }
          }
          onChange={[Function]}
          onSelectMapDefault={[Function]}
          selected={null}
          variableType="query"
        />
      </GridColumn>
    </GridRow>
    <GridRow
      testID="grid--row"
    >
      <GridColumn
        testID="grid--column"
        widthXS={12}
      >
        <FormFooter
          testID="form--footer"
        >
          <Button
            active={false}
            color="danger"
            onClick={[MockFunction]}
            placeIconAfterText={false}
            shape="none"
            size="sm"
            status="default"
            testID="button"
            text="Cancel"
            type="button"
          />
          <Button
            active={false}
            color="primary"
            placeIconAfterText={false}
            shape="none"
            size="sm"
            status="disabled"
            testID="button"
            text="Create"
            type="submit"
          />
        </FormFooter>
      </GridColumn>
    </GridRow>
  </Grid>
</Form>
`;
